pkg/ccl/changefeedccl/changefeed_test.go

```diff
@@ -2419,57 +2431,75 @@ func TestChangefeedLaggingSpanCheckpointing(t *testing.T) {
 	}
 
-	// Should eventually checkpoint all spans around the lagging span
+	// We should eventually checkpoint some spans that are ahead of the highwater.
+	// We'll wait until we have two unique timestamps.
 	testutils.SucceedsSoon(t, func() error {
 		progress := loadProgress()
-		if loadCheckpoint(t, progress) != nil {
+		cp := maps.Collect(loadCheckpoint(t, progress).All())
+		if len(cp) >= 2 {
 			return nil
 		}
-		return errors.New("waiting for checkpoint")
+		return errors.New("waiting for checkpoint with two different timestamps")
 	})
 
 	sqlDB.Exec(t, "PAUSE JOB $1", jobID)
 	waitForJobState(sqlDB, t, jobID, jobs.StatePaused)
 
 	// We expect highwater to be 0 (because we skipped some spans) or exactly cursor
-	// (this is mostly due to racy updates sent from aggregators to the frontier.
+	// (this is mostly due to racy updates sent from aggregators to the frontier).
 	// However, the checkpoint timestamp should be at least at the cursor.
 	progress := loadProgress()
-	require.True(t, progress.GetHighWater().IsEmpty() || *progress.GetHighWater() == cursor,
-		"expected empty highwater or %s,  found %s", cursor, progress.GetHighWater())
+	require.True(t, progress.GetHighWater().IsEmpty() || progress.GetHighWater().Equal(cursor),
+		"expected empty highwater or %s, found %s", cursor, progress.GetHighWater())
 	spanLevelCheckpoint := loadCheckpoint(t, progress)
 	require.NotNil(t, spanLevelCheckpoint)
-	minCheckpointTS := spanLevelCheckpoint.MinTimestamp()
-	require.True(t, cursor.LessEq(minCheckpointTS))
+	require.True(t, cursor.LessEq(spanLevelCheckpoint.MinTimestamp()))
 
+	var rangefeedStarted bool
 	var incorrectCheckpointErr error
 	knobs.FeedKnobs.OnRangeFeedStart = func(spans []kvcoord.SpanTimePair) {
+		rangefeedStarted = true
+
 		setErr := func(stp kvcoord.SpanTimePair, expectedTS hlc.Timestamp) {
 			incorrectCheckpointErr = errors.Newf(
 				"rangefeed for span %s expected to start @%s, started @%s instead",
 				stp.Span, expectedTS, stp.StartAfter)
 		}
 
 		for _, sp := range spans {
-			if laggingSpans.Encloses(sp.Span) {
-				if !sp.StartAfter.Equal(cursor) {
-					setErr(sp, cursor)
+			// NB: It would be nice if we could create a reverse index from
+			// spans to timestamps, but unfortunately roachpb.Span does not
+			// implement the equal operator and so cannot be a map key type.
+			var checkpointTS hlc.Timestamp
+			for ts, spans := range spanLevelCheckpoint.All() {
+				if slices.ContainsFunc(spans, func(s roachpb.Span) bool {
+					return s.Equal(sp.Span)
+				}) {
+					checkpointTS = ts
+					break
+				}
+			}
+			if checkpointTS.IsSet() {
+				if !sp.StartAfter.Equal(checkpointTS) {
+					setErr(sp, checkpointTS)
 				}
 			} else {
-				if !sp.StartAfter.Equal(minCheckpointTS) {
-					setErr(sp, minCheckpointTS)
+				// Any spans not in the checkpoint should be at the cursor.
+				if !sp.StartAfter.Equal(cursor) {
+					setErr(sp, cursor)
 				}
 			}
 		}
 	}
+	knobs.FeedKnobs.ShouldSkipCheckpoint = nil
```

Comment:
is it chill to modify this while the db is running? could lead to data races potentially. i've seen this from other innocuous looking knobs changes.

Reply:
It's done while the changefeed is paused so I think it's okay.